Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -1134,3 +1134,24 @@ void TextureCache::TCacheEntryBase::Zero(u8* ptr)
 		ptr += memory_stride;
 	}
 }
+
+u64 TextureCache::TCacheEntryBase::CalculateHash()
+{
+	u8* ptr = Memory::GetPointer(addr);
+	if (memory_stride == CacheLinesPerRow() * 32)
+	{
+		return GetHash64(ptr, size_in_bytes, g_ActiveConfig.iSafeTextureCache_ColorSamples);
+	}
+	else
+	{
+		u32 blocks = NumBlocksY();
+		u64 temp_hash = size_in_bytes;
+		u32 samples = (g_ActiveConfig.iSafeTextureCache_ColorSamples + blocks - 1) / blocks;
+		for (u32 i = 0; i < blocks; i++)
+		{
+			temp_hash = (temp_hash * 397) ^ GetHash64(ptr, CacheLinesPerRow() * 32, samples);
```

Comment:
It can be hoisted. 

Since CacheLinesPerRow is defined within the same unit of compilation, it will be in-lined and automatically hoisted during optimisation.